rust/reflection/src/safe_buffer.rs

```diff
@@ -0,0 +1,392 @@
+/*
+ * Copyright 2025 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+use crate::r#struct::Struct;
+use crate::reflection_generated::reflection::{Field, Schema};
+use crate::reflection_verifier::verify_with_options;
+use crate::{
+    get_any_field_float, get_any_field_float_in_struct, get_any_field_integer,
+    get_any_field_integer_in_struct, get_any_field_string, get_any_field_string_in_struct,
+    get_any_root, get_field_float, get_field_integer, get_field_string, get_field_struct,
+    get_field_struct_in_struct, get_field_table, get_field_vector, FlatbufferError,
+    FlatbufferResult, ForwardsUOffset,
+};
+use flatbuffers::{Follow, Table, Vector, VerifierOptions};
+use num::traits::float::Float;
+use num::traits::int::PrimInt;
+use num::traits::FromPrimitive;
+use std::collections::HashMap;
+use std::sync::RwLock;
+
+pub struct SafeBuffer<'a> {
+    buf: &'a [u8],
+    schema: &'a Schema<'a>,
+    buf_loc_to_obj_idx: RwLock<HashMap<usize, i32>>,
+}
+
+impl<'a> SafeBuffer<'a> {
+    pub fn new(buf: &'a [u8], schema: &'a Schema) -> FlatbufferResult<Self> {
+        Self::new_with_custom_options(buf, schema, &VerifierOptions::default())
+    }
+
+    pub fn new_with_custom_options(
+        buf: &'a [u8],
+        schema: &'a Schema,
+        opts: &VerifierOptions,
+    ) -> FlatbufferResult<Self> {
+        let mut buf_loc_to_obj_idx = HashMap::new();
+        verify_with_options(&buf, schema, opts, &mut buf_loc_to_obj_idx)?;
+        Ok(SafeBuffer {
+            buf,
+            schema,
+            buf_loc_to_obj_idx: RwLock::new(buf_loc_to_obj_idx),
+        })
+    }
+
+    /// Gets the root table in the buffer.
+    pub fn get_root(&self) -> Table {
+        // SAFETY: the buffer was verified during construction.
+        unsafe { get_any_root(self.buf) }
+    }
+
+    /// Gets all the fields the [table] has. Returns error if the [table] doesn't match the buffer.
```

Comment:
Nit: the "table must match the buffer" constraint sounds like something that would be awesome to check automatically rather than having the user ensure it.

Would it be possible to have something like `SafeTable` (stub name), that keeps a reference to the buffer it refers to? This could give us usage like
```
let buffer: SafeBuffer = SafeBuffer::new(&buf, *schema)?;
let root: SafeTable = buffer.get_root();
root.get_all_fields(); // always correct because SafeBuffer only accesses the matching buffer
```

And there's even a chance we could get rid of `InvalidTableOrStruct` error!